src/cache.rs

```diff
@@ -6,128 +6,80 @@
-        let db_path = data_dir.join("cache.db");
-        if db_path.exists()
-            && !db_path.is_dir()
-            && fs::remove_file(&db_path).is_err()
-        {
-            panic!("Failed to remove old cache files. Please remove {} manually and try again.", db_path.display())
-        }
-
-        CACHE_DATA_PATH
-            .set(data_dir)
-            .expect("cache path can only be set once");
-
-        Persistence::with_backend(&backend(), |_| Ok(()))?;
-
-        Ok(())
-    }
-
-    /// Persist the cache to the internal backend.
-    pub(crate) fn persist(&self) -> Result<(), Error> {
-        let data_dir = CACHE_DATA_PATH.get().expect("cache path must be set");
-        let id_path = data_dir.join("cache.id");
-
-        let persistence_id = Persistence::persist(&backend(), &self.data)?;
-
-        let mut bytes = [0u8; 36];
-        let mut sink = Sink::new(&mut bytes);
-
-        persistence_id.encode(&mut sink);
+    /// Returns a new cache instance.
+    pub(crate) fn new<T: AsRef<Path>>(path: T) -> Result<Self, Error> {
+        let mut opts = Options::default();
+        opts.create_if_missing(true);
+        // After 10 million bytes, sort the cache file and create new one
+        opts.set_write_buffer_size(10_000_000);
 
-        fs::write(id_path, bytes)?;
+        let db = DB::open(&opts, path)?;
+        let write_batch = WriteBatch::default();
 
-        Ok(())
+        Ok(Self { db, write_batch })
     }
 
-    /// Insert a note into the cache at the given block height. If note is
-    /// `None` only the block height gets updated.
+    /// We store (key, Vec<(height, note)>).
     pub(crate) fn insert(
         &mut self,
         psk: PublicSpendKey,
         height: u64,
         note: Option<Note>,
     ) -> Result<(), Error> {
-        if self.data.get(&psk)?.is_none() {
-            self.data.insert(
-                psk,
-                KeyData {
-                    last_height: 0,
-                    notes: BTreeSet::new(),
-                },
-            )?;
+        let psk_bytes = psk.encode_to_vec();
+        let mut data: KeyData;
+
+        if let Some(data_bytes) = self.db.get(&psk_bytes)? {
+            let mut source = Source::new(&data_bytes);
+            // decode key data
+            data = KeyData::decode(&mut source)?;
+
+            if let Some(note) = note {
+                if height > data.last_height {
+                    data.last_height = height;
+                }
+                data.notes.insert(NoteData { height, note });
+            }
+        } else {
+            data = KeyData {
+                last_height: height,
+                notes: BTreeSet::new(),
+            };
+
+            if let Some(note) = note {
+                data.notes.insert(NoteData { height, note });
+            }
         }
 
-        let mut key_data = self.data.get_mut(&psk)?.expect("psk");
+        self.write_batch.put(psk_bytes, data.encode_to_vec());
```

Comment:
Having the whole note set encoded into a singled field/row is something we would like to avoid (mostly to prevent tech-debt that will be raised when we'll start to handle rollback and note tagging).